File: matrix.pl
:- module(matrix,
          [ matrix_new/3,                       % +Type, +Dim, -Matrix

            matrix_type/2,                      % +Matrix, -Type
            matrix_size/2,                      % +Matrix, -NElems

            matrix_set/3,                       % +Matrix, +Position, +Elem
            matrix_get/3,                       % +Matrix, +Position, -Elem

            matrix_set_all/2                    % +Matrix, +Value
          ]).
:- use_module(library(ffi)).

/** <module> Matrix manipulation

@see https://www.dcc.fc.up.pt/~vsc/Yap/documentation.html#matrix
*/

c_define(mptr, *(struct(matrix))).

cpp_const('T_INTS').
cpp_const('T_FLOATS').

:- c_import("#include \"matrix.c\"",
            [ matrix ],
            [ matrix_size_1(int, int, [int]),
              matrix_size_2(int, int, int, [int]),

              matrix_init_1(mptr, int, int),
              matrix_init_2(mptr, int, int, int),

              matrix_type(mptr, [int]) as matrix_typeid,
              matrix_size(mptr, [int]),

              matrix_offset_1(mptr, -int, int, [int]),
              matrix_offset_2(mptr, -int, int, int, [int]),

              matrix_set_all_ints(mptr, int)
            ]).

%!  matrix_new(+Type, +Dim, -Matrix) is det.

matrix_new(Type, [Dim0], Matrix) :-
    !,
    typeid(Type, TypeID),
    matrix_size_1(TypeID, Dim0, Size),
    c_calloc(Matrix, struct(matrix), Size, 1),
    matrix_init_1(Matrix, TypeID, Dim0).
matrix_new(Type, [Dim0, Dim1], Matrix) :-
    typeid(Type, TypeID),
    matrix_size_2(TypeID, Dim0, Dim1, Size),
    c_calloc(Matrix, struct(matrix), Size, 1),
    matrix_init_2(Matrix, TypeID, Dim0, Dim1).

%!  matrix_type(+Matrix, -Type) is det.

matrix_type(Matrix, Type) :-
    matrix_typeid(Matrix, TypeID),
    typeid(Type, TypeID),
    !.

%!  matrix_size(+Matrix, -NElems) is det.
%
%   True when NElems is the number of elements in Matrix.

%!  matrix_set(+Matrix, +Position, +Elem) is det.

matrix_set(Matrix, [Dim0], Elem) :-
    !,
    matrix_offset_1(Matrix, TypeID, Dim0, Offset),
    ctype(TypeID, CType),
    c_store(Matrix, Offset, CType, Elem).
matrix_set(Matrix, [Dim0,Dim1], Elem) :-
    matrix_offset_2(Matrix, TypeID, Dim0, Dim1, Offset),
    ctype(TypeID, CType),
    c_store(Matrix, Offset, CType, Elem).

%!  matrix_get(+Matrix, +Position, -Elem) is det.

matrix_get(Matrix, [Dim0], Elem) :-
    !,
    matrix_offset_1(Matrix, TypeID, Dim0, Offset),
    ctype(TypeID, CType),
    c_load(Matrix, Offset, CType, Elem).
matrix_get(Matrix, [Dim0,Dim1], Elem) :-
    matrix_offset_2(Matrix, TypeID, Dim0, Dim1, Offset),
    ctype(TypeID, CType),
    c_load(Matrix, Offset, CType, Elem).

%!  matrix_set_all(+Matrix, +Elem) is det.
%
%   Set all elements of Matrix to Elem

matrix_set_all(Matrix, Elem) :-
    matrix_typeid(Matrix, TypeID),
    matrix_set_all(TypeID, Matrix, Elem).

matrix_set_all('T_INTS',   Matrix, Elem) :- matrix_set_all_ints(Matrix, Elem).
matrix_set_all('T_FLOATS', Matrix, Elem) :- matrix_set_all_floats(Matrix, Elem).

typeid(ints,   'T_INTS').
typeid(floats, 'T_FLOATS').

ctype('T_INTS',   longlong).
ctype('T_FLOATS', double).

File: matrix.c
#include <stddef.h>
#include <assert.h>

typedef long long m_int;
typedef double    m_float;

#define T_INTS   0
#define T_FLOATS 1

typedef int mtype;

typedef struct matrix
{ int     ndim;
  mtype   type;				/* Type of elements */
  size_t  dim[0];
} matrix;

#define SIZEOF_MHEADER(ndim) \
	((size_t)(&((matrix*)NULL)->dim[ndim]))

static inline size_t
esize(mtype type)
{ switch(type)
  { case T_INTS:   return sizeof(m_int);
    case T_FLOATS: return sizeof(m_float);
    default:	   assert(0);
  }
}

size_t
matrix_size_1(mtype type, size_t dim0)
{ return SIZEOF_MHEADER(1) + esize(type)*dim0;
}

size_t
matrix_size_2(mtype type, size_t dim0, size_t dim1)
{ return SIZEOF_MHEADER(1) + esize(type)*dim0*dim1;
}

void
matrix_init_1(matrix *m, mtype type, size_t dim0)
{ m->ndim   = 1;
  m->type   = type;
  m->dim[0] = dim0;
}

void
matrix_init_2(matrix *m, mtype type, size_t dim0, size_t dim1)
{ m->ndim   = 2;
  m->type   = type;
  m->dim[0] = dim0;
  m->dim[1] = dim1;
}

size_t
matrix_offset_1(const matrix *m, mtype *tp, size_t i0)
{ *tp = m->type;
  return SIZEOF_MHEADER(1) + esize(m->type) * i0;
}

size_t
matrix_offset_2(const matrix *m, mtype *tp, size_t i0, size_t i1)
{ *tp = m->type;
  return SIZEOF_MHEADER(1) + esize(m->type) * (i0*m->dim[0] + i1);
}

mtype
matrix_type(const matrix *m)
{ return m->type;
}
